### DISCUSSION TITLE
Problem on memory consumption

### DISCUSSION THREAD

**Discussion by eanemo** [deepjavalibrary/djl#975] [Q&A]
Hi everyone,

I'm writing some code for implementing a web request that uses a CNN model. Inside this request a need to operate with a DJL NDArray inside a for loop. I have detect that the memory consuption increases about 200Mb after the for loop execution. The main problem is that this memory is not freed at any time. I don't know if this is a problem of my code or a memory leak in the DJL library. The code inside the for loop is the following. Any suggestion is welcome. Thak you in advance.

PD: colors and imgCopy are two NDArrays

```java
    int ii, jj;
    for (int i = 0; i < iidx.size(); i++) {
      ii = iidx.getInt(i);
      jj = jidx.getInt(i);
      String strIdx = Integer.toString(ii) + ", " + Integer.toString(jj) + ", :";

      colors.set(new NDIndex(Integer.toString(i) + ", :"), imgCopy.get(strIdx).div(255));
    }
```

**Comment by frankfliu** [comment]
In DJL, NDArray native memory is released when it's parent NDManager is closed.
In your code, the memory will increase in the loop, and I didn't see you close the NDManager here, if you didn't close it, it will cause leak.

To avoid memory increase in the loop, you do the following:
```
    int ii, jj;
    for (int i = 0; i < iidx.size(); i++) {
      ii = iidx.getInt(i);
      jj = jidx.getInt(i);
      String strIdx = Integer.toString(ii) + ", " + Integer.toString(jj) + ", :";
      NDArray array = imgCopy.get(strIdx);
      NDArray div = array.div(255);
      colors.set(new NDIndex(Integer.toString(i) + ", :"), div);
      div.close();
      array.close();
    }
```

**Reply by eanemo** [reply]
> In DJL, NDArray native memory is released when it's parent NDManager is closed.
> In your code, the memory will increase in the loop, and I didn't see you close the NDManager here, if you didn't close it, it will cause leak.
> 
> To avoid memory increase in the loop, you do the following:
> 
> ```
>     int ii, jj;
>     for (int i = 0; i < iidx.size(); i++) {
>       ii = iidx.getInt(i);
>       jj = jidx.getInt(i);
>       String strIdx = Integer.toString(ii) + ", " + Integer.toString(jj) + ", :";
>       NDArray array = imgCopy.get(strIdx);
>       NDArray div = array.div(255);
>       colors.set(new NDIndex(Integer.toString(i) + ", :"), div);
>       div.close();
>       array.close();
>     }
> ```
First of all, thank you for your response. I have used these temp NDArrays inside the loop closing them at each iteration. I also have closed manually the manager outside the loop. However the memory keeps growing inside the loop. Maybe the memory consumption is about the NDIndex object creation?

**Comment by lanking520** [comment]
Can you simply do 

```
imgCopy.div(255)
```

and then use the outcome as the final ndarray? I am assuming your operation can be done through matrix-wide operation. Can you explain breifly what operation you are trying to achieve here?

**Reply by eanemo** [reply]
> Can you simply do
> 
> ```
> imgCopy.div(255)
> ```
> 
> and then use the outcome as the final ndarray? I am assuming your operation can be done through matrix-wide operation. Can you explain breifly what operation you are trying to achieve here?

First of all, thank you for your response. I am translating some code from Python (Numpy) to DJL. The Python code is as follows:

`colors = img[ind[1, :], ind[0, :], :] / 255.`

where "img" is a numpy array for the image and ind is a numpy array of indices. I have not managed to build an NDIndex similar in DJL, so I am doing it manually in a for loop. Any suggestion?

Thank you.
